### PR TITLE
Allow running `audit` for different architecture.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Same as https://github.com/Homebrew/brew/pull/15056, but for `audit`.

### CODE REVIEW

File: Library/Homebrew/cmd/fetch.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/fetch.rb:25]:
```suggestion
           description: "Download for the given CPU architecture.",
```

File: Library/Homebrew/dev-cmd/audit.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb:36]:
```suggestion
           description: "Audit the given CPU architecture.",
```

**Review by MikeMcQuaid — COMMENTED** [review]
Same as on `fetch`: questions before merging:
- does/will this work on formulae now/later?
- what about running audits on all arches/OS versions/OSs rather than specifying them one by one?

**Comment by reitermarkus** [comment]
> does/will this work on formulae now/later?

For downloading bottles, seems like currently only `--bottle-tag` is considered.

**Comment by MikeMcQuaid** [comment]
> For downloading bottles, seems like currently only `--bottle-tag` is considered.

@reitermarkus I think it would be good for both `brew fetch` and `brew audit` to have a consistent interface for doing this across both formulae and casks.

> I'd rather each arch be tested in a separate CI job.

This pushes too much logic IMO into the cask matrix logic which e.g. can never really be used for formulae. It means you need to reach into the cask (or formulae) to figure out what's needed to test rather than just getting a list of e.g. URLs and figure out what's wrong with them.

**Comment by reitermarkus** [comment]
> I think it would be good for both `brew fetch` and `brew audit` 

I guess it would make sense to replace `--bottle-tag` with `--arch` and `--os` options so they can work for both formulae and casks.

> too much logic IMO into the cask matrix logic

This doesn't push anything new into cask matrix logic. Previously, we would just skip ARM in the matrix completely, now we can at least run `fetch` and `audit`. In either case we have to check the cask contents.

> which e.g. can never really be used for formulae

I don't see what could be reused here for formulae anyways? This problem doesn't exist for formulae since we have ARM runners for `homebrew/core`.

**Comment by MikeMcQuaid** [comment]
> I guess it would make sense to replace `--bottle-tag` with `--arch` and `--os` options so they can work for both formulae and casks.

Yup, that'd be good.

> Previously, we would just skip ARM in the matrix completely, now we can at least run `fetch` and `audit`.

Deciding whether this is done or not is in the matrix logic where it would be better/more useful to have in Homebrew/brew where formulae/other taps can use it too.

> I don't see what could be reused here for formulae anyways? This problem doesn't exist for formulae since we have ARM runners for `homebrew/core`.

Being able to say "download all URLs for all arches/OSs" with a single `brew fetch` invocation would be extremely useful for homebrew/core.

**Comment by reitermarkus** [comment]
> Deciding whether this is done or not is in the matrix logic where it would be better/more useful to have in Homebrew/brew where formulae/other taps can use it too.

Other taps can copy the `ci.yml` workflow file 1:1. Having CI logic in `Homebrew/brew` would just make CI a pain to adjust and debug.

> Being able to say "download all URLs for all arches/OSs" with a single `brew fetch` invocation would be extremely useful for homebrew/core.

The `Utils::Bottles::Tag.from_symbol` method actually supports `:all`, but the way we are handling `--bottle-tag` doesn't involve parsing it with `Utils::Bottles::Tag.from_symbol` at all.

**Comment by MikeMcQuaid** [comment]
> Having CI logic in `Homebrew/brew` would just make CI a pain to adjust and debug.

Not suggesting having all that logic in there, just that more logic in a separate shared repository where it could be tested independently (like `brew test-bot`) and moving more logic (like mentioned here) into Homebrew/brew will be more useful outside the tap.


> The `Utils::Bottles::Tag.from_symbol` method actually supports `:all`, but the way we are handling `--bottle-tag` doesn't involve parsing it with `Utils::Bottles::Tag.from_symbol` at all.

Yeh, `:all` is a very different situation here.

**Comment by MikeMcQuaid** [comment]
> > Yeh, `:all` is a very different situation here.
> 
> What does `:all` do for bottles?

It's a type of bottle that can be used on any OS/version/architecture.

**Comment by reitermarkus** [comment]
Okay, in that case it would download the same bottle regardless of what you pass to `--bottle-tag`?

**Comment by MikeMcQuaid** [comment]
@reitermarkus I think it'll just download an `:all` bottle, regardless of what it is.

**Comment by reitermarkus** [comment]
Can we get this merged to get `audit` in line with `fetch`?

Will take a look at implementing `--os=all --arch=all` in a follow-up PR and making it public.

**Review by MikeMcQuaid — APPROVED** [review]
Yes, that's fair, happy to merge it now assuming future PRs will implement that. Thanks again @reitermarkus!

**Pull request merged**
